### ISSUE TITLE
user can't add_loss w/ input dependence in custom layers in eager mode? (or docs unclear)

### ISSUE THREAD

**Issue by bionicles** [tensorflow/tensorflow#32058]
**System information**
- TensorFlow version (you are using): 2.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Thanks for making tensorflow. 

I want to make custom layers which handle their own losses to write less code 
(ex, layerwise reconstruction + kl divergence losses in stacked autoencoders)
The [docs](https://github.com/tensorflow/tensorflow/blob/a2e398f299e62559118f3e59bd8ef11925cdc449/tensorflow/python/keras/engine/base_layer.py#L1083) indicate this isn't possible in eager mode, which sucks because Eager is default in 2.0... 

I like tf.function / graph mode but frankly it's a different dialect of TF which forces users to waste time translating code into "graph dialect" ... so i want to use Eager, even if performance is worse, it's better than debugging tf.function. 

without modular custom losses, users must write complicated/annoying training loops to apply correct loss function to correct combination of inputs and outputs from a model (and models return arrays so ordering gets complicated... can we name model outputs?)

would it be possible to permit TF users to define module-specific input-dependent losses for custom layers in Eager mode? 

**Will this change the current api? How?**
users could add custom input-dependent losses to custom layers. 
thus, users can dramatically simplify training loops

**Who will benefit with this feature?**
2.0 keras users with custom layers that use custom losses

**Any Other info.**
I want to make these "coder" bricks into Layers which handle reconstruction + KL divergences in a modular way
```
def get_sensor_and_actuator(agent, in_spec):
    if in_spec.rank is 3:
        sensor = use_image_sensor(agent)
        actuator = use_image_actuator(agent)
    elif in_spec.rank is 2 and in_spec.shape[1] is None:
        sensor = use_ragged_sensor(agent, in_spec)
        actuator = use_ragged_actuator(agent, in_spec)
    else:
        sensor = use_resizer(agent.code_spec.shape)
        actuator = use_resizer(in_spec.shape)
    return sensor, actuator


def use_coder(agent, in_spec):
    log('use_coder', in_spec, color="blue")
    normalizer = norm = use_norm()

    if in_spec.rank is 3:
        h, w = get_hw(in_spec.shape)
        hw = [h, w]

        def resize_then_norm(x):
            x = tf.image.resize(x, hw)
            return norm(x)
        normalizer = resize_then_norm

    coordinator = L.Lambda(concat_coords)
    sensor, actuator = get_sensor_and_actuator(agent, in_spec)

    def call(x):
        normie = normalizer(x)
        normie_w_coords = coordinator(normie)
        code = sensor(normie_w_coords)
        reconstruction = actuator(code)
        return normie, code, reconstruction
    return call
```
alas, since i can't use input-dependent modular losses, I need to
1. design my model to return a bunch of extra outputs
2. keep track of the order of those outputs. which one is a code? which one is a reconstruction?
3. unpack the outputs of the model to organize codes, reconstructions, normalized inputs, and "actual" desired outputs
4. organize pairs of these normalized inputs and reconstructions
5. loop over the pairs and apply a loss function
6. append error terms to a list of losses.

Is there some reason we cannot add a loss function to the layers themselves and avoid all this BS?

this would REALLY simplify modular agents... please advise!

**Comment by bionicles** [comment]
bump @ymodak @oanush @fchollet @dynamicwebpaige is this a docs issue or a feature request?

TLDR: to simplify code, I want to make a "coder" layer which computes the reconstruction loss in the call method; but i'm using Eager, and the docs indicate this is impossible ... 

here's what I want to do:
```
class Coder(Layer):
    def __init__(self, in_spec, out_spec):
          super(Coder, self).__init__()
          self.norm = InstanceNormalization()
          self.sensor = Dense(out_spec.size)
          self.actuator = Dense(in_spec.size)

     def call(self, x):
            normie = self.norm(x)
            code = self.sensor(normie)
            reconstruction = self.actuator(code)
            **self.add_loss(MSLE(normie, reconstruction))**  # <--- does this work?
            return code
```
note: i'm using the functional method to build a keras model. The goal of the coder is to learn a latent representation of the inputs. It's much more complicated to put this loss in the training loop

**Comment by bionicles** [comment]
```
import tensorflow as tf

from nature import Resizer, ConcatCoords2D, AllAttention, Norm
from tools import concat_coords

K = tf.keras
L = K.layers

L1 = 1e-3
L2 = 1e-3

class Predictor(L.Layer):
    """generates inputs and returns a surprise value"""

    def __init__(self, agent, in_spec):
        super(Predictor, self).__init__()
        self.batch_size = agent.batch_size
        self.out_spec = agent.code_spec
        self.in_spec = in_spec

    def build(self, shape):
        self.regularize = L.ActivityRegularization(l1=L1, l2=L2)
        self.resizer = Resizer(self.out_spec.shape)
        self.coordinator = ConcatCoords2D()
        self.attention = AllAttention(self.batch_size)
        self.subtract = L.Subtract()
        self.noise_norm = Norm()
        self.x_norm = Norm()
        self.built = True

    @tf.function
    def call(self, x):
        noise = tf.random.normal(tf.shape(x))
        noise = self.coordinator(noise)
        noise = self.noise_norm(noise)
        p = self.attention(noise)
        x = self.coordinator(x)
        p = tf.reshape(p, tf.shape(x))
        x = self.x_norm(x)
        surprise = self.subtract([x, p])
        **self.add_loss(surprise)  
        # self.add_loss(K.losses.MSLE(x, p))
        # surprise = self.regularize(surprise) 
        # both fail here, in both eager and tf.function** 
        return self.resizer(surprise)

    def compute_output_shape(self, shape):
        return self.out_spec
```
cc @omalleyt12 @ymodak @oanush @fchollet 
trying this custom layer for predictive coding, but it gives errors like:

> TypeError: An op outside of the function building code is being passed
> a "Graph" tensor. It is possible to have Graph tensors
> leak out of the function building context by including a
> tf.init_scope in your function building code.
> For example, the following function will fail:
>   @tf.function
>   def has_init_scope():
>     my_constant = tf.constant(1.)
>     with tf.init_scope():
>       added = my_constant * 2
> The graph tensor has name: dense_1/ActivityRegularizer/add_1:0
^^^ if training loop is eager

with @tf.function on training loop:

>     InaccessibleTensorError: The tensor 'Tensor("resizer/dense/ActivityRegularizer/Cast:0", shape=(), dtype=float32)' cannot be accessed here: it is defined in another function or code block. Use return values, explicit Python locals or TensorFlow collections to access it. Defined in: FuncGraph(name=call, id=139631001707744); accessed from: FuncGraph(name=reduce_reduce_body, id=139631001171336).

here is the training loop:
```
    @tf.function
    def run_data_session(self, data, model, loss_fn):
        for step, (image, label) in data.enumerate():
            with tf.GradientTape() as tape:
                loss = loss_fn(label, model(image))
            gradients = tape.gradient(
                **[loss] + model.losses**, model.trainable_variables)
            self.optimizer.apply_gradients(
                zip(gradients, model.trainable_variables))
            tf.print(step, loss)
```
 `model.losses` triggers these errors, if we don't attempt to get those losses, then it works

**Comment by bionicles** [comment]
hey, it works if tf.function is on the training loop and not on the predictor call

**Comment by cottrell** [comment]
I'm hitting similar things but using tfp as well so not sure if there is a good pattern to avoid this.

**Comment by cmcneil** [comment]
If I'm not mistaken, this also means that this tutorial does not run in eager mode: https://www.tensorflow.org/tutorials/generative/deepdream
(Hits InaccessibleTensorErrors as well)

**Comment by giovp** [comment]
Having the same issue, it seems related to #32477, any news on this?

**Comment by omalleyt12** [comment]
@bionicles Thanks for the issue!

Yes, this is supported, the issue is indeed the `tf.function` wrapping the Layer.call. You can remove this, as `Model.fit` runs inside a `tf.function` anyway, and for custom training loops it's more performant to wrap your whole training step or training loop in a `tf.function`

We're in the process of updating documentation, this is one "gotcha" we should document better

**Comment by Andreas5739738** [comment]
Trying to briefly summarize this issue, which I also ran into:
- Layer.add_loss(...) does not work in eager mode
- it works in graph mode, but _not_ if Layer.call is decorated with tf.function

Should we consider this a satisfying state of things? Having code that runs in graph mode but fails in eager mode is problematic for a number of reasons. E.g., it makes debugging and testing complicated  (you cannot even call Layer.call in a unit test since it fails in eager mode and cannot be decorated with tf.function).
